### PR TITLE
Feature (#1659): Toggle visible trees

### PR DESCRIPTION
**Pull request by richard-fine**
As per issue #1659, add a new view menu toggle for controlling visibility of trees (separately from other scenery).

### CODE REVIEW

**Comment by AaronVanGeffen** [comment]
This is making the trees completely invisible, right? I ask because the other options give the objects a transparent shade instead (except the peep option).

**Comment by richard-fine** [comment]
Yes, this makes them completely invisible.

**Comment by tupaschoal** [comment]
Should be pretty clear, but can you add a before/after screenshot  and one showing how the toggle looks like?

Do we go with transparent or invisible @AaronVanGeffen ?

**Comment by AaronVanGeffen** [comment]
Personally, I would prefer it to be consistent, that is transparent. I would welcome input from other team members, though.

**Review by tupaschoal — COMMENTED** [review]
To me it looks like the original issue wanted trees to not be hidden along with scenery. If that's the case, I'd go with transparent, to keep the consistency.

But if the goal is not to split it off from the "toggle scenery", then invisible seems alright.

File: distribution/changelog.txt [hunk 1]
```diff
@@ -9,6 +9,7 @@
 - Feature: [#12885] Add SmallSceneryElement.quadrant to the plugin API.
 - Feature: [#12886] Make all scenery placement and remove actions available to the plugin API.
 - Feature: [#2350, #12922] Add snow, heavy snow and blizzard to weather types.
+- Feature: [#1659] Add 'Invisible Trees' view toggle option.
```
[the remark this answers is not in the document]
Comment (AaronVanGeffen) [reply distribution/changelog.txt:12]: (Probably best to update it after 0.3.1 has been released.)

**Comment by ZehMatt** [comment]
I think transparency would be the better choice to keep things consistent for hidden things.

**Comment by richard-fine** [comment]
> Should be pretty clear, but can you add a before/after screenshot and one showing how the toggle looks like?

Here's a screen recording of the feature in action:

ezgif com-video-to-gif

File: src/openrct2-ui/windows/TopToolbar.cpp [hunk 2]
```diff
@@ -134,12 +134,13 @@ enum TOP_TOOLBAR_VIEW_MENU_DDIDX {
...
     DDIDX_INVISIBLE_SUPPORTS = 7,
     DDIDX_INVISIBLE_PEEPS = 8,
-    DDIDX_LAND_HEIGHTS = 10,
-    DDIDX_TRACK_HEIGHTS = 11,
-    DDIDX_PATH_HEIGHTS = 12,
+    DDIDX_INVISIBLE_TREES = 9,
+    DDIDX_LAND_HEIGHTS = 11,
+    DDIDX_TRACK_HEIGHTS = 12,
+    DDIDX_PATH_HEIGHTS = 13,
     // 13 is a separator
```
[the remark this answers is not in the document]
Comment (richard-fine) [reply src/openrct2-ui/windows/TopToolbar.cpp]: Whoops! Good catch.

**Comment by richard-fine** [comment]
Regarding transparent vs hidden - as I see it you could argue things either way. Transparent is consistent with 'See-Through Rides', 'See-Through Scenery', etc. Hidden is consistent with 'Invisible People' and 'Invisible Supports.'

I don't have a very strong opinion either way, I just implemented this as it was described in the feature request, but it's not like the request has a lot of discussion on it. If we can reach a consensus that it should be transparent instead, I can change this PR to do that instead.

**Comment by IntelOrca** [comment]
I am wondering whether we want to separate the flags out to (hidden / not hidden) and (transparent / partial).

**Comment by richard-fine** [comment]
A sliding scale of "Visible --- transparent --- hidden" might be nice, not just for trees but for the other things as well. I'm not sure how to do the UI for that, though.

**Comment by ocalhoun6** [comment]
@richard-fine 
You could have no UI for it at all ... just have the extra options normally only accessible via the console ... and then let people develop plugins (or develop one yourself) that activate the extra hidden view modes.

**Comment by AaronVanGeffen** [comment]
A granular UI for this is plugin material, imo. Up to us to expose flags for it, though, as @IntelOrca suggested.

**Comment by richard-fine** [comment]
Ah, yes, that makes a lot of sense. So I think I have a few options here:

- I could just introduce a second viewport flag for VIEWPORT_FLAG_SEETHROUGH_TREES and make it work, then we try and figure out which of the two flags should have a menu item (if either).
- I could introduce viewport flags for all the things that are currently in that menu, e.g. VIEWPORT_FLAG_SEETHROUGH_SUPPORTS, VIEWPORT_FLAG_INVISIBLE_RIDES, etc. I think we'd need another 9 flags to cover everything currently in the menu; we have space for 11 right now.
- I could pull the visibility settings out of the viewport flags entirely, and store them in a separate pair of bitvectors, making it easier for us to add more visibility classes in the future - for example I think being able to control walls/rooftops separately from the rest of scenery might be helpful.

What do you reckon?

**Comment by IntelOrca** [comment]
I think there should be two fields:
* viewport flags (categories of what to hide)
* transparency flags (invisible or transparent for each category)

Whether we allow each viewport to have different transparency flags as well as different viewport flags is a different matter.

**Comment by richard-fine** [comment]
@IntelOrca something like this?

```
#include <bitset>

enum class ViewportVisibilityGroup
{
    BaseGround,
    VerticalFaces,
    Rides,
    Scenery,
    Supports,
    Peeps,
    Paths,
    Trees,

    Count
};

class VisibilityState
{
    std::bitset<static_cast<size_t>(ViewportVisibilityGroup::Count)> _visible;
    std::bitset<static_cast<size_t>(ViewportVisibilityGroup::Count)> _transparent;

public:
    bool IsVisible(ViewportVisibilityGroup group)
    {
        return _visible.test(static_cast<size_t>(group));
    }

    bool IsTransparent(ViewportVisibilityGroup group)
    {
        return _transparent.test(static_cast<size_t>(group));
    }

    void SetVisible(ViewportVisibilityGroup group, bool visible)
    {
        _visible.set(static_cast<size_t>(group), visible);
    }

    void SetTransparent(ViewportVisibilityGroup group, bool transparent)
    {
        _visible.set(static_cast<size_t>(group), transparent);
    }
};
```

**Comment by Gymnasiast** [comment]
We discussed this in the team chat. We prefer if this PR just added the option to make trees see-through (in line with other scenery, rides and footpaths). The see-through vs. invisible stuff, which could use more discussion anyway, is best left for another PR.

**Comment by Gymnasiast** [comment]
Are you still working on this?

**Comment by richard-fine** [comment]
I don't know when I will get back to it. If someone else wants to pick it up, they can go ahead.

**Comment by Gymnasiast** [comment]
Closing it for now. Let us know if you want to continue working on it.

**Pull request closed without merging**